### ISSUE TITLE
Add support for lazy loading large result sets

### ISSUE THREAD

**Issue by chicks** [chicks/sugarcrm#15]
Problem:
Base#find currently returns ALL results - this can result in slow performance or memory limits when querying large datasets.

Solution: 
Update #find to return a result set enumerator that utilizes SugarCRM's entry_list => offset functionality.  This is essentially how database cursors work - it should track the total number of results, the current offset, and the results per page / max_results.

The fog gem does this really well - we can steal some examples from there.

**Comment by davidsulc** [comment]
Yes, this is definitely a very useful feature, as it will allow iterating on large record collections.

However, it seems SugarCRM has a bug when query record slices.

`SugarCRM::Contact.all(:order_by => 'last_name', :limit => 10)`

returns

Alpha
Bravo
Charlie
Delta
Echo
Foxtrot
Golf
Hotel
India
Juliet

`SugarCRM::Contact.all(:order_by => 'last_name', :limit => 3, :offset => 6)`

returns

Golf
Hotel
India

`SugarCRM::Contact.all(:order_by => 'last_name', :limit => 6, :offset => 3)`

returns

Golf
Hotel
India

(N.b. results are also wrong if queried via `SugarCRM.connection.get_entry_list`)

In other words, it seems SugarCRM is incorrectly doing something like
`limit = [offset, limit].min`

`offset = [offset, limit].max`

**Comment by chicks** [comment]
Ooooh... interesting.  I can submit a bug for this.

**Comment by davidsulc** [comment]
No lazy loading, but I changed `Base#find_by_sql` to work around timeouts. See commit https://github.com/chicks/sugarcrm/commit/9c23fc723af93f8d28fdaa516af3d67482a2a138

This also seems to fix the `:limit` > `:offset` bug.

**Comment by chicks** [comment]
I'm about halfway done with this feature.

**Comment by chicks** [comment]
Also, I think the :limit > :offset bug might be unique to 5.5.  Have you considered upgrading?

**Comment by davidsulc** [comment]
`:limit` and `:offset` parameters still don't behave correctly (on more recent Sugar versions) when sent directly to Sugar via API (on v. 6.1.1, build 5512):

```
SugarCRM.connection.get_entry_list('Accounts', "1=1", :order_by => 'name', :limit => 10).each{|a| puts a.name }
puts '---------------------------------'
SugarCRM.connection.get_entry_list('Accounts', "1=1", :order_by => 'name', :limit => 3, :offset => 6).each{|a| a.name }
puts '---------------------------------'
SugarCRM.connection.get_entry_list('Accounts', "1=1", :order_by => 'name', :limit => 6, :offset => 3).each{|a| puts a.name }
```

Output:

```
1: A.G. Parr PLC
2: AB Drivers Limited
3: Aim Capital Inc
4: Airline Maintenance Co
5: Anytime Air Support Inc
6: Chandler Logistics Inc
7: Complete Holding
8: Davenport Investing
9: Gifted Holdings AG
10: Gifted Holdings AG
---------------------------------
1: Complete Holding
2: Davenport Investing
3: Gifted Holdings AG
---------------------------------
1: Complete Holding
2: Davenport Investing
3: Gifted Holdings AG
```

This is fixed, however, when going via the gem:

```
SugarCRM::Account.all(:order_by => 'name', :limit => 10).each{|a| puts a.name }
puts '---------------------------------'
SugarCRM::Account.all(:order_by => 'name', :limit => 3, :offset => 6).each{|a| puts a.name }
puts '---------------------------------'
SugarCRM::Account.all(:order_by => 'name', :limit => 6, :offset => 3).each{|a| puts a.name }
```

Output:

```
1: A.G. Parr PLC
2: AB Drivers Limited
3: Aim Capital Inc
4: Airline Maintenance Co
5: Anytime Air Support Inc
6: Chandler Logistics Inc
7: Complete Holding
8: Davenport Investing
9: Gifted Holdings AG
10: Gifted Holdings AG
---------------------------------
1: Complete Holding
2: Davenport Investing
3: Gifted Holdings AG
---------------------------------
1: Airline Maintenance Co
2: Anytime Air Support Inc
3: Chandler Logistics Inc
4: Complete Holding
5: Davenport Investing
6: Gifted Holdings AG
```

**Comment by davidsulc** [comment]
Can we close this?
1. Users can pass blocks to finders instead (bypassing memory and performance issues on large datasets):
   
   SugarCRM::Accounts.all{|a| puts a.name}
2. :limit and :offset bug is invisible to gem users, due to workaround in find_by_sql

**Comment by chicks** [comment]
Closing for now.
